Organization Chart

bkrith opened the pull request:
Fixes #2730 

## Overview

Basic organization chart for contacts.
Used d3 and d3-org-chart libraries.

## Requirements
Have contacts with Org Manager field.
And, the root contact with Org Manager set to "Head".
![image](https://user-images.githubusercontent.com/5513284/166169266-28c049bb-0d26-443f-aff0-f2ce9c3c1526.png)

### Add field orgmanager in contact
![image](https://user-images.githubusercontent.com/5513284/166169080-d6c134c4-fd0f-402a-b4de-69f05491b3e5.png)

### Organization chart
![image](https://user-images.githubusercontent.com/5513284/166169101-52ec4749-2416-4eec-9a49-f0c6364d042c.png)

### Highlight path to root
![image](https://user-images.githubusercontent.com/5513284/166169127-1a75023a-2a69-44d5-9e30-26e8c0469380.png)

## Todo

- [x] What is the element in the top right, are there different org charts to switch between? Is this something we want to directly have in version 1, and if so how many charts are there usually? (Cause I would assume it would be just one?)
- [x] The background color would look better as color-main-background instead of grey
- [ ] Currently there are a lot of inline styles. This results in the app not working with dark theme, and it also being difficult to change. Best is to use the variables supplied by Nextcloud :)
- [x] Currently everything is divs without semantics, which is not accessible. For example, the name of people need to be `h3`, the title and the info lines below `p`
- [x] The placeholder icon for people without avatar is not Nextcloud standard – if you use NcAvatar you will automatically get placeholders https://nextcloud-vue-components.netlify.app/#/Components/NcAvatar
- [x] The blue bar on top of each person looks nice – but does not seem to be taken from `var(--color-primary)` – would be better to do that. Also the whitespace above that bar can be removed.
- [x] The highlight path too root should also be color-primary, and to make it more obvious you can use a border of 2px. Also the border is not around the existing box, there’s another container around it?
- [x] The default border should be `border: 1px solid var(--color-border)` using the variable
- [x] You can use the border radius value `var(--border-radius-large)` on the containers
- [ ] Manages/oversees uses custom small text sizes which are not accessible. Just use the default font size / no need to change it :)
- [x] For people who do not manage or oversee anyone, that part does not need to be shown at all
- [x] The icons next to manages/oversees introduce visual noise and are not needed, can be removed :)
- [ ] BUG new avatars are transparent https://github.com/nextcloud/nextcloud-vue/pull/3272
  ![Bildschirmfoto vom 2022-09-19 14-45-01](https://user-images.githubusercontent.com/1374172/191020401-9743c2e5-3cee-414d-be96-feb19f96169a.png)
- [ ] BUG missing handling of deleted managers
  ![Bildschirmfoto vom 2022-09-19 14-54-16](https://user-images.githubusercontent.com/1374172/191021927-46fd37a2-7613-483e-b546-f00d6827923e.png)
- [ ] BUG cyclic references (alice is bob's manager, bob is alice's manager) show blank chart page with empty list of charts
  ![Bildschirmfoto vom 2022-09-19 14-55-41](https://user-images.githubusercontent.com/1374172/191022254-d626c61c-21d2-47e7-a6da-17025c16e1b0.png)
- [ ] BUG graph element is not responsive. Enlarging the window causes partly hidden graphs.
  ![Bildschirmfoto vom 2022-09-19 15-01-24](https://user-images.githubusercontent.com/1374172/191023325-6198243c-7184-499e-96dd-fb2bfac85e9d.png) 
  ![Bildschirmfoto vom 2022-09-19 15-01-36](https://user-images.githubusercontent.com/1374172/191023351-4113e947-6493-4837-b260-ccd1154c2fa5.png)
- [ ] Graph card link says `https://localhost/apps/contacts/chart/All%20contacts` but click on avatar gets me to `https://localhost/apps/contacts/All%20contacts/4cee56a8-c43c-4e4a-bcba-55026faf1687~contacts`. This means a ctrl+click (or right click and open in new tab) leads to a different page than an ordinary click
- [x] To research (for the lack of reproducibility at the moment): handling of shared address books and cross-addressbook references in general. I guess managers should only be assignable within the same addressbook so that we can assure there are no broken links.
- [x] Show more levels by default -> all levels?
- [ ] Order chart dropdown by alphabet
- [x] Hide chart dropdown when there is only one option
- [ ] Expand button is rather small
- [x] Height of card header is not even -> some cards have a bigger height

ChristophWurst commented:
This is quite an interesting feature. Before we jump into any details, could you please open a feature request at https://github.com/nextcloud/contacts/issues/new?assignees=&labels=enhancement%2C0.+Needs+triage&template=feature_request_form.yml so we can have a discussion about what we want to have and how it should look? I don't want to bloat the pull request with brainstorming comments :)

Cheers

florisvangeel commented:
This PR is the result of an idea proposal at [NextGov 2022](https://eventornado.com/submission/organisational-chart-of-employee-contacts) Hackathon

**Abstract**:
In contacts there is one special circle with the company directory.
This helper app will create two relations from this contact to its manager and side reporter.
With that information the company directory can get a live organisational chart of all the people.
Using an open source library 
https://www.npmjs.com/package/d3-org-chart

there might be some UX improvements needed. This challenge comes down to defining the datasets and building the first prototype within nextcloud.

As a result every team can be organised.

Presentation: https://docs.google.com/presentation/d/1r_J-nop5r-cgDtSQpHPNbwJ5igaCmftc/
Demo video:  https://www.youtube.com/watch?v=vM_2yXlhbak

R7JANV1 commented:
A great extension!

But many companies and organizations have several supervisors...

DineshRefex commented:
pls help me to build org chart

DineshRefex reviewed (commented):
not working for me

bkrith commented:
I'm sorry for the delay.. I was busy with my dayjob..

@DineshRefex maybe it was a confusion with HEAD value which needed for the top member(root) of chart.

Please try again I have removed this value(exists but it's not necessary anymore) and also added support for multiple charts.

bkrith commented:
Dropdown at the right top with all available charts generated from contacts.
For now the name of chart is the org field of root contact and the fullName.
![image](https://user-images.githubusercontent.com/5513284/184670110-d2e47712-ca83-4261-bbc8-c6b866795c94.png)

DineshRefex commented:
Thank you so much, I'll work on it and send you the result. Once again
thank you for spending your time with me.

On Mon, Aug 15, 2022 at 9:26 PM Vassilis ***@***.***> wrote:

> Dropdown at the right top with all available charts generated from
> contacts.
> For now the name of chart is the org field of root contact and the
> fullName.
> [image: image]
> <https://user-images.githubusercontent.com/5513284/184670110-d2e47712-ca83-4261-bbc8-c6b866795c94.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nextcloud/contacts/pull/2726#issuecomment-1215227042>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZH7Y7VOYGYNNI2BQC2VKQDVZJSERANCNFSM5U2SKATQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 
*The information contained in this communication is intended solely for the 
use of the individual or entity to whom it is addressed and others 
authorized to receive it. **Please note that as the recipient, it is your 
responsibility to check the email for malicious content or software. **It 
may contain confidential or legally privileged information. If you are not 
the intended recipient you are hereby notified that any disclosure, 
copying, distribution or taking any action in reliance on the contents of 
this information is strictly prohibited and may be unlawful. If you have 
received this communication in error, please notify us immediately by 
responding to this email.The firm is neither liable for the proper and 
complete transmission of the information contained in this communication 
nor for any delay in its receipt.*

ChristophWurst commented on l10n/el.js:
```diff
@@ -240,6 +240,8 @@ OC.L10N.register(
     "Pager" : "Βομβητής",
     "Voice" : "Ομιλία",
     "Car" : "Αμάξι",
+    "Organization Chart": "Οργανόγραμμα",
```
translations are handled via Transifex, not directly in our code. You can revert these changes.

ChristophWurst commented on src/components/OrgChart.vue:
```diff
@@ -0,0 +1,164 @@
+					const imgEl = `<img src="${d.data.image || 'data:image/jpeg;base64,/9j/4AAQSkZJRgABAgAAZABkAAD/7AARRHVja3kAAQAEAAAAZAAA/+4ADkFkb2JlAGTAAAAAAf/bAIQAAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQICAgICAgICAgICAwMDAwMDAwMDAwEBAQEBAQECAQECAgIBAgIDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMD/8AAEQgAjABkAwERAAIRAQMRAf/EAG0AAQACAgMBAQAAAAAAAAAAAAAJCgYIAwQHAgUBAQAAAAA ...
+
+					return `
+						<div style="padding-top:30px;background-color:none;margin-left:1px;height:${d.height}px;border-radius:2px;overflow:visible">
```
is there any chance for d3 to work with Vue components instead of string concatenation so that we can safely prevent user content injection into the DOM?

ChristophWurst commented on src/components/OrgChart.vue:
```diff
@@ -0,0 +1,164 @@
+									<div style="color:#404040;font-size:16px;margin-top:4px"> ${d.data.title}</div>
+								</div> 
+								<div style="display:flex;justify-content:space-between;padding-left:15px;padding-right:15px;">
+									<div> Manages:  ${d.data._directSubordinates} 👤</div>  
```
l10n

ChristophWurst commented on src/models/contact.js:
```diff
@@ -364,6 +364,31 @@ export default class Contact {
+	 * @memberof Contact
+	 */
+	get orgManager() {
+		return this.firstIfArray(this.vCard.getFirstPropertyValue('orgmanager'))
```
is this possibly a standardized property? I'm not able to find references of ORGMANAGER in the vcard specs. If it's not, should be prefix with `X-` @skjnldsv?

skjnldsv replied:
Yep, if not official, prefix is mandatory

CarlSchwan replied:
If you use X-ManagersName, it will be compatible with /kalendar, better than nothing 😜

A similar concept exist in google eith manager and assistant name but they don't even use vcards

jancborchardt replied:
We already have "Relationship to you" with "Manager", "Colleague" and "Assistant", is that not a standard of sorts we can use here too?
![image](https://user-images.githubusercontent.com/925062/185460651-f1641592-8453-4ce0-be02-765c30fe7637.png)

bkrith replied:
I just pushed commit with all the requested changes/bugs/etc.
Currently I'm using X-ManagersName.. Relationship field is like first person property and I'm not sure(never used) if is anything else except an info.. except if extend/change it..

ChristophWurst replied:
*Relationship to you* is not helpful because it only stores the type of relation and is always relative to the addressbook owner. E.g. in your addressbook there might be a contact for me and it says *Colleague*. Nothing more, nothing less.

`X-ManagersName` seems a lot more appropriate. The only issue we just found out is that KOrganizer stores the *name* of the other person for this field, not the UID. So it is not a reference to another contact. It is just a value. And this value is prone for conflicts in larger organizations where names are not unique.

I'll see if we can put together some kind of hybrid solution where we store the UID for our purposes but also X-ManagersName for interoperability.

ChristophWurst replied:
We will go with something like `X-MANAGERSNAME;UID=abc-def-123:John`

ChristophWurst reviewed (commented):
It works

![Bildschirmfoto vom 2022-08-18 14-39-11](https://user-images.githubusercontent.com/1374172/185396767-07ae80f4-250c-42c5-9ca5-893476708105.png)

ChristophWurst commented:
- [ ] BUG it should probably not be possible to be someones own org manager (self reference), or any kind of loops in general

ChristophWurst commented:
- [ ] BUG if you don't set any org managers the chart page is simply blank. This could be filled with a placeholder text. Or we hide the page if no contact has a manager set.

ChristophWurst commented:
- [ ] BUG: top right shows *null chart*?!

ChristophWurst commented:
- [ ] BUG: contact cards show `null` – is this a missing title or something? in any case, should be hidden if there is no meaningful text to display

ChristophWurst commented:
- [ ] BUG: clicking a contact card triggers a full page reload. It should be possible to navigate with a Vue `$router.push` https://router.vuejs.org/guide/essentials/navigation.html.

jancborchardt reviewed (changes requested):
Nice, cool feature! :)
Some points regarding design, as it would be nice to adjust this to the [Nextcloud design guidelines](https://docs.nextcloud.com/server/latest/developer_manual/design/index.html) :)

Edit: moved todos up https://github.com/nextcloud/contacts/pull/2726#issue-1222328379

Do you have anything else @nimishavijay?

bkrith commented:
Thank you all so much for your feedback!!
I have already moved the node template in vue component with avatar compatibility, etc. and not inline text (it was a little bit tricky..)
I will fix all the issues/bugs soon..

@jancborchardt about the first question.. some people complain that want multiple organization charts from the same contacts.. I will make it better with label to be obvious and i will remove it in case we have only one chart/tree/organization.
the most issues are from the fast copy paste html template inline text.. will be fixed with vue component, avatar, etc instead of inline text/html/css as you describe.

Again, thank you for your feedback/guidelines.

Any comment, feedback and guideline is always welcome!

jancborchardt commented:
@bkrith awesome, sounds good! :) Let me know if there’s a new version to review – screenshots make it very easy to do a quick design review.

bkrith commented:
@jancborchardt I have committed requested changes.. I will upload tonight some screenshots.. I'm using X-ManagersName(debatable..)

bkrith commented:
![image](https://user-images.githubusercontent.com/5513284/189520298-f6f37cbb-ad2e-4aab-9fa6-8ee5f8a9f014.png)

![image](https://user-images.githubusercontent.com/5513284/189520339-1c030480-0882-4391-bbfb-9cca0632eab8.png)

ChristophWurst commented:
Locally updated to latest `main`. After https://github.com/nextcloud/contacts/pull/2917 three imports of `@nextcloud/vue` have to be adjusted.

ChristophWurst commented:
- [ ] BUG I can no longer select a manager

![Bildschirmfoto vom 2022-09-16 17-57-26](https://user-images.githubusercontent.com/1374172/190680710-99504eb2-639f-42db-ad63-8b476cbb8bca.png)

ChristophWurst commented:
Today's state with the new interface

![Bildschirmfoto vom 2022-09-19 15-13-04](https://user-images.githubusercontent.com/1374172/191026689-33b93721-74f1-49fa-8601-1ac15f7eabd2.png)

ChristophWurst commented:
@bkrith I have checked my previously commented bugs against the current state. Great job. Those are all resolved. I haven't had the chance to also check Jan's design feedback. Since you don't have permission to edit other people's comments and markdown checklist, I moved the todos up into your PR description so you can check the items and mark everything that's already done.

I will also add any bugs to the todo list as I encounter them.

ChristophWurst commented on src/components/ChartTemplate.vue:
```diff
@@ -0,0 +1,53 @@
+<template>
+	<div class="node-container">
+		<div class="inner-box">
+			<a href="" @click.prevent.stop="onAvatarClick(data.nodeId)">
```
this should be a Vue router link

ChristophWurst commented on css/orgChart.scss:
```diff
@@ -0,0 +1,89 @@
+.orgchart-container {
```
please break this up into scoped scss that is put into the Vue SFCs

ChristophWurst commented:
I'd say let's get the remaining bugs fixed, then we can merge a first version of this feature. Anything else can be done as follow-ups.

ChristophWurst commented on src/components/AppNavigation/RootNavigation.vue:
```diff
@@ -42,6 +42,16 @@
+					name: 'chart',
+					params: { selectedChart: GROUP_ALL_CONTACTS },
+				}"
+				icon="icon-category-monitoring" />
```
icon classes are deprecated. this should use a material icon

ChristophWurst commented on src/components/AppNavigation/RootNavigation.vue:
```diff
@@ -281,6 +292,11 @@ export default {
 
+		// check if any contact has manager, if not then is no need for organization chart menu
+		existChart() {
+			return !!Object.keys(this.contacts).filter(key => this.contacts[key].managersName).length
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some as a lightweight alternative to `.filter(...).length`

ChristophWurst commented on src/utils/chartUtils.js:
```diff
@@ -0,0 +1,29 @@
+import { GROUP_ALL_CONTACTS } from '../models/constants.ts'
+
+export const getChart = (list, parent, nodes = []) => {
+	if (!nodes.length) nodes.push(list.find(node => node.nodeId === parent))
```
please use blocks for if and no one-line ifs for better readability

ChristophWurst commented:
@bkrith if you allow us to push commits to your branch we can fix some of the reported problems right away

bkrith commented:
Thank you! I'm too busy right now.. I'm moving on new apartment and I have overloaded in my day job.

ChristophWurst commented:
https://github.com/nextcloud/contacts/pull/2954 is ready for review and test

DineshRefex commented:
Hi there ***@***.***, I have tried my level best but once I have
done. It asking me to upgrade, after that no improvement it remains the
same.*

*I'm using 24 version of nextcloud server. Will that be ok, or else I have
to swap the version. Help me in stucked with orgchat. *





On Thu, 18 Aug, 2022, 6:09 pm Christoph Wurst, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> It works
>
> [image: Bildschirmfoto vom 2022-08-18 14-39-11]
> <https://user-images.githubusercontent.com/1374172/185396767-07ae80f4-250c-42c5-9ca5-893476708105.png>
> ------------------------------
>
> In l10n/el.js
> <https://github.com/nextcloud/contacts/pull/2726#discussion_r949055924>:
>
> > @@ -240,6 +240,8 @@ OC.L10N.register(
>
>      "Pager" : "Βομβητής",
>
>      "Voice" : "Ομιλία",
>
>      "Car" : "Αμάξι",
>
> +    "Organization Chart": "Οργανόγραμμα",
>
>
> translations are handled via Transifex, not directly in our code. You can
> revert these changes.
> ------------------------------
>
> In src/components/OrgChart.vue
> <https://github.com/nextcloud/contacts/pull/2726#discussion_r949058395>:
>
> > +				this.chartReference = new OrgChart()
>
> +			}
>
> +			this.chartReference
>
> +				.container(this.$refs.svgElementContainer) // node or css selector
>
> +				.data(data)
>
> +				.nodeWidth((d) => 250)
>
> +				.initialZoom(0.8)
>
> +				.nodeHeight((d) => 175)
>
> +				.childrenMargin((d) => 40)
>
> +				.compactMarginBetween((d) => 15)
>
> +				.compactMarginPair((d) => 80)
>
> +				.nodeContent(function(d, i, arr, state) {
>
> +					const imgEl = `<img src="${d.data.image || 'data:image/jpeg;base64,/9j/4AAQSkZJRgABAgAAZABkAAD/7AARRHVja3kAAQAEAAAAZAAA/+4ADkFkb2JlAGTAAAAAAf/bAIQAAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQICAgICAgICAgICAwMDAwMDAwMDAwEBAQEBAQECAQECAgIBAgIDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMD/8AAEQgAjABkAwERAAIRAQMRAf/EAG0AAQACAgMBAQAAAAAAAAAAAAAJCgYIAwQHAgUBAQAAAAAAAAAAAAAAAAAAAAAQAAEEAgEDBAEEAgIDAAAAAAIAAQMEBQYHESESExQVCAkxIhYXQTJxIzMkNBEBAAAAAAAAAAAAAAAAAAAAAP/aAAwDAQACEQMRAD8Av8ICAgICAgICAgICAgICAgICAg+SIQEjMhAAFyIidhERFupERP0YRFm6u7/og69S/RvgUtG5UuxgfgclSxDZAD6MXgRwmYifiTP0fv0dB2kBAQEBAQEBAQEEQv2r/I++t5PJcf8A1+PH5DJUJZqOa5Jtww5HG1LkJyRT1NSx84SUcqdeQej37DS1SdnaKKUXGZBD/unJPIPI148lve6bNttwjIxkz+Zv5KOByd38KdaxOdalCPX9scIBGLdmZmQYxjMrlMLciyGHyV/E34H6w3sZcsULkL9WLrFZqyRTRv5Cz9ibuyCQj68/kW5V4zyFHDcpXMhyloZnDBZmyMwT7xhYG6RnbxedsmB5sowdzKtkZDeZxYQsQdSJwnt0feNV5I1XDbrpWZq57W89VG3jsjUJ/Eh6uEtexEbDNUvVJhKKeCURlhlEgMWJnZBliAgICAgICCNL8jv2QucX6LS4o1K5JU3Hkqhaky1+ubx2cJo4ySUbhQGJCcVzZLQyVIjHr4QQ2f8AU/TJgr+ICAgIJFPx3fYu7xdyjU4wz18v4ByfkYcfHBPI/t8Hu9kYqmDy1di8hiHMnHHj7TN4sfnBIZdK7M4WHEBAQEBAQEFXL7n79Y5E+y/KuUkm9Wlg9js6ViQEhOCLH6aT6+71iFyYoLt+lPa6s7sRWHduzszBq4gICAg5a889WeG1Wlkgs1pY5688JlHLDPCYyRSxSC7EEkcgs4u3dnZBbm4h3X+xuLOO97IhebbNN13OXBBhEYsjfxdabJ12EGYBetkCljdh7M49uyD0VAQEBAQEFQTkozl5G3+SQykkk3XajkkMnMzM87fIjMidyIiJ+ru/d3QYUgICAgILQH0blkl+qXDZSyHITYPLRMUhkZNHDtOehhjZyd3YIoYxEW/QRFmbsyDbBAQEBAQEFSrn3ASatzhy9r5xvGOL5J3OvXF/P91J9gvy4+VvUOSTxnoyRm3kRP0Lu7ug8kQEBAQEFq76o6+WsfW3hTFSRtFK/HuvZWaJh8Hjn2GoOwThIPiLtMMuUdj6t18+vXr+qDYJAQEBAQEFcz8kuhSah9lMpn44vDHci69gtprkH/iG7Uq/xrKwt/lpns4RrBt3/wDpZ/8APRg0CQEBAQZZoepX993bUdIxYk+Q23ZMLrtRxBz9KXL5CvRawbCz9IazTvIZP2EBd36Mzugt80KNXF0KWNoxDBSx1StRpwD18YatSEK9eIeru/jHFGzN/wAIO2gICAgICCOf8lHC1nkbhipvmEpna2LiW5bzM8cIFJPY03KRQQ7QIADfufGnTq3yIn6RVqs/RupIK86AgICCUb8YfCE+1ck5TmbL0y/j3HUM+M1+SaJ3gv7pmaRwSlCRsUUv8fwVo5JG/wBo57lYxfqyCepAQEBAQEBBw2K8FuCeraghs1bMMlezWsRhNBYgmAo5oJ4ZBKOWGWMnEhJnEhd2duiCs394+B8DwHzZLgtS9YNT2zA1d0wdCbyNsGGQyeWx13AxWSZnsV6F3FGcHXqYVpogMjMXMg05QEGW6DqN7kDedO0bGm0V/cNnwWs1ZyApI602bydbHDamEO/oVfcepI/ZmAXd3Zm6oLYvGHGuqcQ6Nr/Hul0ipYHXqnt4HmIJLt6zKZT3spkpwjiGzkslbkOaY2ERcy6CIgwiwZ8gICAgICAgIK9f5PtwxuxfYbHYLGzQzno+h4bCZcoy83hzN/IZfYJapkzuHWDGZWo7s3cTMmfu3RgjjQEHrHA+14vReauKtwzZvFhdc3/Vcrl52/Wti6uZqHftszCTm9Sp5yePbz8fHq3XqwW1YpY5o45oZAlhlAJYpYjGSOSOQWIJIzF3EwMXZ2dndnZ0H2gICAgICDgtWq1KtPcu2IKlSrDJYtWrU0detWghFzlnnnlIIoYYgF3IidhFm6u6CKn7VfkZ13VqmS0XgK/V2XbJGkp3+QIgit6xr3Xyjl/jzyMcGyZYW/0nZix8TuxMVh+oCEHeVymSzmTyOazN61lMvl71rJZTJXp5LN3IZC9Odm5dt2JSKWezZsSkZmTu5ETu6DoICAgkl+o337z/AA7HjuPuVCyG18ZQjFTxOSj62tj0mAegRQ1PUMSy+uwD29oRetXBm9uXiLQEE8OmbxqHImAp7Ro+x4naMBfFir5PEW47UPn4iR17Ai7TU7kPkzSQTDHNEXYxF+yDKkBAQa985/Z/iD6+UQk33YHkzlmD3GN03AhDk9syMLkQhPHjSs1oaFOQgJhsXJq1cyAhEyIXFBFJyT+VHk/NS2anGGma5pONLyjgyeeebati6N5MNmMHfHYKmZ9WJ4jrXBF26eZN3cNDuSOf+Z+XXIeReR9m2SoUgzfES3Bx+vDMPXxmi1vERY/AQzD17EFYSZu3VB4+gICAgICDNNI5H33jXKfM6BuGw6hki8GmsYHK28e1uOMvMIL8EEg1sjVYu7xThJG/+RdBu3o/5NPsdrD14dlLUOQqcbiM5Z3AhicqcLO3aC/q82GqRz+LdPUlq2OrO7kJF3Qb88Q/k24b3m1Uw/IeJynFWWtGEIZC7YHYNQKY3YAabN06tS/jmkkfuVikFaIe5zszO6CRX5rDfD/yH5bGfAfH/LfOe/q/D/Fe3918n8n6vsvj/a/9nrefp+n+7r07oKfmx7HntvzmT2baMvfz2fzNo7uUy+TsyW712zIzM8k08ruTsICwgLdBABYRZhZmYPxUBAQEBAQEBAQEBB6P/b3JX9a/1B/MMx/XHzHzn8W9dvY++6+p6fqeHuvjvdf+z7P1Pae7/wC/0/W/eg84QEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQf/Z'}" alt="${d.data.fullName}" style="margin-top:-30px;margin-left:${d.width / 2 - 30}px;border-radius:100px;border:1px solid lightgray;width:60px;height:60px;background-color: white" />`
>
> +
>
> +					return `
>
> +						<div style="padding-top:30px;background-color:none;margin-left:1px;height:${d.height}px;border-radius:2px;overflow:visible">
>
>
> is there any chance for d3 to work with Vue components instead of string
> concatenation so that we can safely prevent user content injection into the
> DOM?
> ------------------------------
>
> In src/components/OrgChart.vue
> <https://github.com/nextcloud/contacts/pull/2726#discussion_r949058497>:
>
> > +				.nodeContent(function(d, i, arr, state) {
>
> +					const imgEl = `<img src="${d.data.image || 'data:image/jpeg;base64,/9j/4AAQSkZJRgABAgAAZABkAAD/7AARRHVja3kAAQAEAAAAZAAA/+4ADkFkb2JlAGTAAAAAAf/bAIQAAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQICAgICAgICAgICAwMDAwMDAwMDAwEBAQEBAQECAQECAgIBAgIDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMD/8AAEQgAjABkAwERAAIRAQMRAf/EAG0AAQACAgMBAQAAAAAAAAAAAAAJCgYIAwQHAgUBAQAAAAAAAAAAAAAAAAAAAAAQAAEEAgEDBAEEAgIDAAAAAAIAAQMEBQYHESESExQVCAkxIhYXQTJxIzMkNBEBAAAAAAAAAAAAAAAAAAAAAP/aAAwDAQACEQMRAD8Av8ICAgICAgICAgICAgICAgICAg+SIQEjMhAAFyIidhERFupERP0YRFm6u7/og69S/RvgUtG5UuxgfgclSxDZAD6MXgRwmYifiTP0fv0dB2kBAQEBAQEBAQEEQv2r/I++t5PJcf8A1+PH5DJUJZqOa5Jtww5HG1LkJyRT1NSx84SUcqdeQej37DS1SdnaKKUXGZBD/unJPIPI148lve6bNttwjIxkz+Zv5KOByd38KdaxOdalCPX9scIBGLdmZmQYxjMrlMLciyGHyV/E34H6w3sZcsULkL9WLrFZqyRTRv5Cz9ibuyCQj68/kW5V4zyFHDcpXMhyloZnDBZmyMwT7xhYG6RnbxedsmB5sowdzKtkZDeZxYQsQdSJwnt0feNV5I1XDbrpWZq57W89VG3jsjUJ/Eh6uEtexEbDNUvVJhKKeCURlhlEgMWJnZBliAgICAgICCNL8jv2QucX6LS4o1K5JU3Hkqhaky1+ubx2cJo4ySUbhQGJCcVzZLQyVIjHr4QQ2f8AU/TJgr+ICAgIJFPx3fYu7xdyjU4wz18v4ByfkYcfHBPI/t8Hu9kYqmDy1di8hiHMnHHj7TN4sfnBIZdK7M4WHEBAQEBAQEFXL7n79Y5E+y/KuUkm9Wlg9js6ViQEhOCLH6aT6+71iFyYoLt+lPa6s7sRWHduzszBq4gICAg5a889WeG1Wlkgs1pY5688JlHLDPCYyRSxSC7EEkcgs4u3dnZBbm4h3X+xuLOO97IhebbNN13OXBBhEYsjfxdabJ12EGYBetkCljdh7M49uyD0VAQEBAQEFQTkozl5G3+SQykkk3XajkkMnMzM87fIjMidyIiJ+ru/d3QYUgICAgILQH0blkl+qXDZSyHITYPLRMUhkZNHDtOehhjZyd3YIoYxEW/QRFmbsyDbBAQEBAQEFSrn3ASatzhy9r5xvGOL5J3OvXF/P91J9gvy4+VvUOSTxnoyRm3kRP0Lu7ug8kQEBAQEFq76o6+WsfW3hTFSRtFK/HuvZWaJh8Hjn2GoOwThIPiLtMMuUdj6t18+vXr+qDYJAQEBAQEFcz8kuhSah9lMpn44vDHci69gtprkH/iG7Uq/xrKwt/lpns4RrBt3/wDpZ/8APRg0CQEBAQZZoepX993bUdIxYk+Q23ZMLrtRxBz9KXL5CvRawbCz9IazTvIZP2EBd36Mzugt80KNXF0KWNoxDBSx1StRpwD18YatSEK9eIeru/jHFGzN/wAIO2gICAgICCOf8lHC1nkbhipvmEpna2LiW5bzM8cIFJPY03KRQQ7QIADfufGnTq3yIn6RVqs/RupIK86AgICCUb8YfCE+1ck5TmbL0y/j3HUM+M1+SaJ3gv7pmaRwSlCRsUUv8fwVo5JG/wBo57lYxfqyCepAQEBAQEBBw2K8FuCeraghs1bMMlezWsRhNBYgmAo5oJ4ZBKOWGWMnEhJnEhd2duiCs394+B8DwHzZLgtS9YNT2zA1d0wdCbyNsGGQyeWx13AxWSZnsV6F3FGcHXqYVpogMjMXMg05QEGW6DqN7kDedO0bGm0V/cNnwWs1ZyApI602bydbHDamEO/oVfcepI/ZmAXd3Zm6oLYvGHGuqcQ6Nr/Hul0ipYHXqnt4HmIJLt6zKZT3spkpwjiGzkslbkOaY2ERcy6CIgwiwZ8gICAgICAgIK9f5PtwxuxfYbHYLGzQzno+h4bCZcoy83hzN/IZfYJapkzuHWDGZWo7s3cTMmfu3RgjjQEHrHA+14vReauKtwzZvFhdc3/Vcrl52/Wti6uZqHftszCTm9Sp5yePbz8fHq3XqwW1YpY5o45oZAlhlAJYpYjGSOSOQWIJIzF3EwMXZ2dndnZ0H2gICAgICDgtWq1KtPcu2IKlSrDJYtWrU0detWghFzlnnnlIIoYYgF3IidhFm6u6CKn7VfkZ13VqmS0XgK/V2XbJGkp3+QIgit6xr3Xyjl/jzyMcGyZYW/0nZix8TuxMVh+oCEHeVymSzmTyOazN61lMvl71rJZTJXp5LN3IZC9Odm5dt2JSKWezZsSkZmTu5ETu6DoICAgkl+o337z/AA7HjuPuVCyG18ZQjFTxOSj62tj0mAegRQ1PUMSy+uwD29oRetXBm9uXiLQEE8OmbxqHImAp7Ro+x4naMBfFir5PEW47UPn4iR17Ai7TU7kPkzSQTDHNEXYxF+yDKkBAQa985/Z/iD6+UQk33YHkzlmD3GN03AhDk9syMLkQhPHjSs1oaFOQgJhsXJq1cyAhEyIXFBFJyT+VHk/NS2anGGma5pONLyjgyeeebati6N5MNmMHfHYKmZ9WJ4jrXBF26eZN3cNDuSOf+Z+XXIeReR9m2SoUgzfES3Bx+vDMPXxmi1vERY/AQzD17EFYSZu3VB4+gICAgICDNNI5H33jXKfM6BuGw6hki8GmsYHK28e1uOMvMIL8EEg1sjVYu7xThJG/+RdBu3o/5NPsdrD14dlLUOQqcbiM5Z3AhicqcLO3aC/q82GqRz+LdPUlq2OrO7kJF3Qb88Q/k24b3m1Uw/IeJynFWWtGEIZC7YHYNQKY3YAabN06tS/jmkkfuVikFaIe5zszO6CRX5rDfD/yH5bGfAfH/LfOe/q/D/Fe3918n8n6vsvj/a/9nrefp+n+7r07oKfmx7HntvzmT2baMvfz2fzNo7uUy+TsyW712zIzM8k08ruTsICwgLdBABYRZhZmYPxUBAQEBAQEBAQEBB6P/b3JX9a/1B/MMx/XHzHzn8W9dvY++6+p6fqeHuvjvdf+z7P1Pae7/wC/0/W/eg84QEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQf/Z'}" alt="${d.data.fullName}" style="margin-top:-30px;margin-left:${d.width / 2 - 30}px;border-radius:100px;border:1px solid lightgray;width:60px;height:60px;background-color: white" />`
>
> +
>
> +					return `
>
> +						<div style="padding-top:30px;background-color:none;margin-left:1px;height:${d.height}px;border-radius:2px;overflow:visible">
>
> +							<div class="inner-box" style="height:${d.height - 32}px;padding-top:0px;background-color:white;border:1px solid lightgray;border-radius:5px">
>
> +								<a href="${d.data.link}">
>
> +									${imgEl}
>
> +								</a>
>
> +								<div style="margin-top:-27px;background-color:#3AB6E3;height:10px;width:${d.width - 2}px;border-radius:1px" />
>
> +								<div style="padding:20px; padding-top:35px;text-align:center">
>
> +									<div style="color:#111672;font-size:16px;font-weight:bold"> ${d.data.fullName}</div>
>
> +									<div style="color:#404040;font-size:16px;margin-top:4px"> ${d.data.title}</div>
>
> +								</div>
>
> +								<div style="display:flex;justify-content:space-between;padding-left:15px;padding-right:15px;">
>
> +									<div> Manages:  ${d.data._directSubordinates} 👤</div>
>
>
> l10n
> ------------------------------
>
> In src/models/contact.js
> <https://github.com/nextcloud/contacts/pull/2726#discussion_r949061158>:
>
> > @@ -364,6 +364,31 @@ export default class Contact {
>
>  		this.vCard.updatePropertyWithValue('org', org)
>
>  	}
>
>
>
> +	/**
>
> +	 * Return the first orgManager
>
> +	 *
>
> +	 * @readonly
>
> +	 * @memberof Contact
>
> +	 */
>
> +	get orgManager() {
>
> +		return this.firstIfArray(this.vCard.getFirstPropertyValue('orgmanager'))
>
>
> is this possibly a standardized property? I'm not able to find references
> of ORGMANAGER in the vcard specs. If it's not, should be prefix with X-
> @skjnldsv <https://github.com/skjnldsv>?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nextcloud/contacts/pull/2726#pullrequestreview-1077238491>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZH7Y7W55DEMZAZU5G4O5ZLVZYVIXANCNFSM5U2SKATQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 
*The information contained in this communication is intended solely for the 
use of the individual or entity to whom it is addressed and others 
authorized to receive it. **Please note that as the recipient, it is your 
responsibility to check the email for malicious content or software. **It 
may contain confidential or legally privileged information. If you are not 
the intended recipient you are hereby notified that any disclosure, 
copying, distribution or taking any action in reliance on the contents of 
this information is strictly prohibited and may be unlawful. If you have 
received this communication in error, please notify us immediately by 
responding to this email.The firm is neither liable for the proper and 
complete transmission of the information contained in this communication 
nor for any delay in its receipt.*
